### PR TITLE
avocado.core.loader: introduce --loaders to avocado run [v1]

### PR DESCRIPTION
**Pull request by ldoktor**
Hello guys,

First commit improves the output when plugin fails to initialize. Then I refactor the inner_runner and move it from a separate plugin (which was in reality already quite intrusively spread across loaders) directly into the file loader. Second to last commit makes it possible to query mapping on uninitialized plugin, which is essential for some features of --loaders.

The last big commit adds the new functionality. It allows specifying either loader plugins (`@vt`, `@file`), test type (`VT`, `SIMPLE`, `INSTRUMENTED`) and for some plugins it allows even extra params (`@file:/bin/echo`, `INNER_RUNNER:/bin/echo -e -n`). There is a hardcoded keyword `DEFAULT` which injects the unused loader plugins in that position (this means `DEFAULT` => `@vt @file`, but `DEFAULT @vt` => `@file @vt`. It doesn't repeat already used loaders when specified by plugin name. It does repeat the ones specified by test type: `SIMPLE @DEFAULT` => `@file:SIMPLE @vt @file`).

The _test type_ is not enforced by the loader_proxy. First I implemented it there, but it did not scale and the only plugin which implements multiple test types is the `file_loader`, therefor I made it optional via the `extra_params`. I still feel it's a bit fragile, because we must trust loader plugin developers to implement such feature for loaders with multiple test types. On the other hand this is more flexible and currently covers all scenarios.

Note in RFC it was mentioned we should support:

* `avocado run --loader avocado a.py --loader simple b.py --loader wrapped:"/usr/bin/test-runner -x -y -z" c.py --loader auto /test-dir/`: multiple test sets, each with a different explicit test loader.

This was not implemented as argparse doesn't support mixed arguments (at least the last time I checked, which was while working on the multiplexer, where I wanted to support similar feature).

v0: https://github.com/avocado-framework/avocado/pull/801

Changes:

    v1: Documentation
    v1: Selftests
    v1: Support to query for loaders (separate commit)
    v1: Adjusted messages and the usage of assert

### CODE REVIEW

**Comment by lmr** [comment]
That's some nice work you've got there, @ldoktor. Please change a few things in the documentation and I think we'll be good to go (I'm re-testing).

File: docs/source/Loaders.rst [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+also supply ``DEFAULT``, which injects into that position all the remaining
+unused loaders.
+
+Examples which lists what tests would be executed with various options
+(with avocado-vt installed)::
+
+    $ avocado run --loaders ?
+    $ avocado run --loaders @file:?
+    $ avocado run passtest boot this_does_not_exists /bin/echo
+        > INSTRUMENTED passtest.py:PassTest.test
```
Comment (clebergnu) [inline docs/source/Loaders.rst]: When I run this example I get:

```
./scripts/avocado run passtest boot this_does_not_exists /bin/echo
Avocado crashed: TypeError: __init__() takes exactly 2 arguments (3 given)
Traceback (most recent call last):

  File "/home/cleber/src/avocado/avocado/avocado/core/job.py", line 451, in _run
    test_suite = self._make_test_suite(urls)

  File "/home/cleber/src/avocado/avocado/avocado/core/job.py", line 260, in _make_test_suite
    loader.loader.load_plugins(self.args)

  File "/home/cleber/src/avocado/avocado/avocado/core/loader.py", line 144, in load_plugins
    self._initialized_plugins.append(plugin(args, extra_params))

TypeError: __init__() takes exactly 2 arguments (3 given)
```

Comment (lmr) [inline docs/source/Loaders.rst]: This PR requires https://github.com/avocado-framework/avocado-vt/pull/166 applied to avocado-vt for things to work.

Comment (clebergnu) [inline docs/source/Loaders.rst]: Oh thanks! My bad!

**Pull request closed without merging**